### PR TITLE
Neutron physics

### PR DESCRIPTION
**Pull request by amandalund**
This PR adds a script for validating neutron transport in OpenMC (thanks for getting this started, @paulromano!). There is also a new option to generate cross section libraries for OpenMC on the fly from MCNP data.

### CODE REVIEW

File: neutron-physics/model.py [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+        elif A == 242 and m == 1:
+            m = 0
+
+        if re.match('(7[0-4]c)|(8[0-6]c)', suffix):
+            suffix = f'71{suffix[1]}nc'
+
+        return f'{1000000*m + 1000*Z + A}.{suffix}'
+
+    def _create_library(self):
+        """Convert the ENDF ACE data from the MCNP distribution into an HDF5
```
Comment (paulromano) [inline neutron-physics/model.py]: Convert the ACE data...

File: neutron-physics/model.py [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+        self.energy = energy
+        self.particles = particles
+        if not re.match('(7[0-4]c)|(8[0-6]c)|(71[0-6]nc)', suffix):
+            msg = f'Unsupported MCNP cross section suffix {suffix}.'
+            raise ValueError(msg)
+        self.suffix = suffix
+        if library is not None and not Path(library).is_dir():
+            msg = f'{library} is not a directory.'
+            raise ValueError(msg)
+        self.library = library
```
Comment (paulromano) [inline neutron-physics/model.py]: To simplify matters later, consider
```Python
if library is not None:
    self.library = Path(library)
```

File: neutron-physics/model.py [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+                try:
+                    table = openmc.data.ace.get_table(path, zaid)
+                except ValueError:
+                    pass
+        else:
+            # Get the table from the ENDF/B-VII.1 neutron ACE files
+            Z, A, m = openmc.data.zam(self.nuclide)
+            element = openmc.data.ATOMIC_SYMBOL[Z]
+            szax = self.szax(self.nuclide, self.suffix)
+            path = Path(self.library) / f'endf71x/{element}/{szax}'
```
Comment (paulromano) [inline neutron-physics/model.py]:
```Python
path = self.library / 'endf71x' / f'{element}' / f'{szax}'
```

File: neutron-physics/model.py [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+            table = openmc.data.ace.get_table(path, szax)
+
+        # Convert cross section data
+        data = openmc.data.IncidentNeutron.from_ace(table, 'mcnp')
+
+        # Export HDF5 file
+        os.makedirs('openmc', exist_ok=True)
+        h5_file = Path(f'openmc/{data.name}.h5')
```
Comment (paulromano) [inline neutron-physics/model.py]:
```Python
h5_file = Path('openmc') / f'{data.name}.h5'
```

File: neutron-physics/model.py [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+        data.export_to_hdf5(h5_file, 'w')
+
+        # Register with library and write cross_sections.xml
+        data_lib = openmc.data.DataLibrary()
+        data_lib.register_file(h5_file)
+        data_lib.export_to_xml(Path('openmc/cross_sections.xml'))
```
Comment (paulromano) [inline neutron-physics/model.py]:
```Python
Path('openmc') / 'cross_sections.xml'
```

File: neutron-physics/model.py [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+        # Directory from which openmc is run
+        os.makedirs('openmc', exist_ok=True)
+        
+        # Define material
+        mat = openmc.Material()
+        mat.add_nuclide(self.nuclide, 1.0)
+        mat.set_density('g/cm3', self.density)
+        materials = openmc.Materials([mat])
+        if self.library is not None:
+            xs_path = Path('openmc/cross_sections.xml').resolve()
```
Comment (paulromano) [inline neutron-physics/model.py]:
```Python
(Path('openmc') / 'cross_sections.xml').resolve()
```

Comment (paulromano) [reply neutron-physics/model.py]: I think you get the idea (use `/` operator as opposed to embedding it in a string)

File: photon-physics/model.py [hunk 7]
```diff
@@ -37,20 +45,118 @@ class Model(object):
...
+    def __init__(self, material, density, elements, energy, particles,
+                 electron_treatment, photon_library=None, name=None):
         self.material = material
         self.density = density
         self.elements = elements
         self.energy = energy
-        self.electron_treatment = electron_treatment
         self.particles = particles
+        self.electron_treatment = electron_treatment
+        if photon_library is not None and not Path(photon_library).is_dir:
```
Comment (paulromano) [inline photon-physics/model.py]: `is_dir()`?

File: photon-physics/validate.py [hunk 8]
```diff
@@ -7,20 +7,25 @@
...
                     help='Energy of the source in eV')
+parser.add_argument('-p', '--particles', type=int, default=1000000,
+                    help='Number of source particles')
 parser.add_argument('-t', '--electron-treatment', choices=('ttb', 'led'),
                     default='ttb', help='Whether to use local energy'
                     'deposition or thick-target bremsstrahlung treatment '
                     'for electrons and positrons.')
-parser.add_argument('-n', '--particles', type=int, default=100000,
-                    help='Number of source particles')
+parser.add_argument('-k', '--photon-library', type=str,
```
Comment (paulromano) [inline photon-physics/validate.py]: Not `-l`?

**Review by paulromano — CHANGES_REQUESTED** [review]
Thanks for wrapping this up! How did the results look?

**Comment by amandalund** [comment]
So far so good! I haven't looked at everything yet, but I'll let you know if I see an discrepancies.

**Comment by paulromano** [comment]
@amandalund One more thing needed here is a temperature specification on the OpenMC model. For example, if I try to run with 71c data from MCNP, I get an error when running OpenMC because it is looking for data at 293.6 K by default (and the 71c data is at 600 K). Can you add that in? The easiest thing to do is to just change the global default temperature:
```Python
settings = openmc.Settings()
settings.temperature = {'default': value}
```
where value is the temperature in K that the data is at which you can get via:
```Python
data = openmc.data.IncidentNeutron.from_ace(table, 'mcnp')
temperature = data.kTs[0] / openmc.data.K_BOLZTMANN
```

File: neutron-physics/model.py [hunk 9]
```diff
@@ -0,0 +1,387 @@
...
+            settings.temperature = {'default': self._temperature}
+        settings.source = source
+        settings.particles = self.particles
+        settings.run_mode = 'fixed source'
+        settings.batches = 1
+        settings.create_fission_neutrons = False
+        settings.export_to_xml(Path('openmc') / 'settings.xml')
+ 
+        # Define tallies
+        energy_bins = np.logspace(3, np.log10(self.energy), 500)
```
Comment (paulromano) [inline neutron-physics/model.py]: For neutrons, we need to go much lower than 1 keV. I would probably tally all the way down to 1e-5 eV as that is generally the lowest point in the data.

File: neutron-physics/model.py [hunk 10]
```diff
@@ -0,0 +1,387 @@
...
+        """
+        # Directory from which MCNP will be run
+        os.makedirs('mcnp', exist_ok=True)
+
+        # Create the problem description
+        lines = ['Point source in infinite geometry']
+ 
+        # Create the cell cards: material 1 inside sphere, void outside
+        lines.append('c --- Cell cards ---')
+        lines.append(f'1 1 -{self.density} -1 imp:n=1')
```
Comment (paulromano) [inline neutron-physics/model.py]: Apparently you also need `tmp=value` here where value is kT in units of MeV. MCNP uses a free gas scattering treatment that defaults to T=293.6K and applies that in the absence of any temperature specification. The following should do:
```Python
kT = self._temperature * openmc.data.K_BOLTZMANN * 1e-6
...
lines.append(f'1 1 -{self.density} -1 imp:n=1 tmp={kT}')
```

File: neutron-physics/model.py [hunk 11]
```diff
@@ -0,0 +1,387 @@
...
+ 
+        # Set grid and tick marks
+        ax1.tick_params(axis='both', which='both', direction='in', length=10)
+        ax1.grid(b=False, axis='both', which='both')
+        ax2.tick_params(axis='y', which='both', right=False)
+        ax2.grid(b=True, which='both', axis='both', alpha=0.5, linestyle='--')
+ 
+        # Set axes labels and limits
+        ax1.set_xlim([1.e-11, self.energy_mev])
+        ax1.set_xlabel('Energy (MeV)', size=12)
```
Comment (paulromano) [inline neutron-physics/model.py]: I'd prefer units of eV :)

File: neutron-physics/model.py [hunk 12]
```diff
@@ -0,0 +1,387 @@
...
+        ax1.set_ylabel('Spectrum', size=12)
+        ax1.legend()
+        ax2.set_ylabel("Relative error", size=12)
+        title = f'{self.nuclide}, {self.energy_mev} MeV Source'
+        plt.title(title)
+ 
+        # Save plot
+        os.makedirs('plots', exist_ok=True)
+        if self.name is None:
+            name = f'{self.nuclide}.png'
```
Comment (paulromano) [inline neutron-physics/model.py]: Also include energy in the filename as we do for photons? Maybe temperature too?

File: neutron-physics/model.py (hunk 10 above)
Comment (amandalund) [reply neutron-physics/model.py]: Ah, thanks for catching this.

File: neutron-physics/model.py (hunk 12 above)
Comment (amandalund) [reply neutron-physics/model.py]: Yeah, I'll make that the default.

File: neutron-physics/model.py (hunk 10 above)
Comment (paulromano) [reply neutron-physics/model.py]: Yeah, I looked at some results for 600 K and they looked way off. Knew something had to be wrong!

**Review by paulromano — APPROVED** [review]

**Comment by paulromano** [comment]
You are my #1 superstar @amandalund. Thanks!

**Pull request merged**